### PR TITLE
8335091: NMT: VMATree reserve_mapping and commit_mapping APIs need MEMFLAGS while un/-committing API has no MEMFLAGS arg

### PR DESCRIPTION
**Pull request by afshin-zafari**
In committing a region, it is not mandatory to provide a MEMFLAGS flag where the committed region inherits the flag from the main region it resides in.
In un-committing there is no need to a  MEMFLAGS at all. 
The `register_mapping` API of the VMATree *requires* a MEMFLAGS (via metadata arg) in both of these two operations. To do the flag inheriting, it is possible to copy the flag of the left node in the tree to the newly inserted ones. 

An optional bool arg (default is false) is added to VMATree API to copy the existing flag of the left node to the new nodes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335091](https://bugs.openjdk.org/browse/JDK-8335091): NMT: VMATree reserve_mapping and commit_mapping APIs need MEMFLAGS while un/-committing API has no MEMFLAGS arg (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20330/head:pull/20330` \
`$ git checkout pull/20330`

Update a local copy of the PR: \
`$ git checkout pull/20330` \
`$ git pull https://git.openjdk.org/jdk.git pull/20330/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20330`

View PR using the GUI difftool: \
`$ git pr show -t 20330`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20330.diff">https://git.openjdk.org/jdk/pull/20330.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20330#issuecomment-2250193372)

### CODE REVIEW

**Review by jdksjolen — CHANGES_REQUESTED** [review]
Hi,

The general idea of allowing for an overriding strategy in order to let `register_mapping` support some form of the current `uncommit` API is fine. Some notes:

1. The new feature is currently unused. Please implement a user, for example `VMATree::uncommit_mapping` and provide tests for this method also.
2.  What happens if we have the following:

```c++
tree.reserve_mapping(0, 10, rd1);
tree.reserve_mapping(10000, 10, rd2, true);
```

Seems like this will have the second reservation have `mtTest`, no? Maybe the `copy_flag` should only be used for when the starting address is a direct match?

Thanks,
Johan

File: test/hotspot/gtest/nmt/test_vmatree.cpp
Comment (gerard-ziemski) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:178]: When does this happen, that we reserve an overlapping region with a different NMT flag?

Comment (afshin-zafari) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:178]: Here in this test, I used a different flag to verify that the new flag is not used for the new nodes of the tree. This is testing even if a flag is given but the arg is true, it will be ignored.

**Comment by afshin-zafari** [comment]
> 1. The new feature is currently unused. Please implement a user, for example `VMATree::uncommit_mapping` and provide tests for this method also.
> 2. What happens if we have the following:

@jdksjolen,  in the latest commit I have fixed these issues.

File: test/hotspot/gtest/nmt/test_vmatree.cpp
Comment (jdksjolen) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:178]: > When does this happen, that we reserve an overlapping region with a different NMT flag?

Uncommitting is essentially the same as reserving. We don't do this in the codebase right now, because that'd make NMT crash, but there's nothing weird about the pattern itself.

**Comment by jdksjolen** [comment]
Alright, I think this is OK.

One thing that worries me a bit is the fact that we take the `leq` value without checking if it's equal to the `A` address. This means that if we only have a range 0-100 with mtTest and I reserve with `copy_flag` at 1000-1100, then that range will receive `mtTest`. That's a bit spooky. Perhaps the `copy_flag` should only be used if there's a range starting at exactly that address? That seems less surprising to me.

**Comment by afshin-zafari** [comment]
> One thing that worries me a bit is the fact that we take the `leq` value without checking if it's equal to the `A` address. This means that if we only have a range 0-100 with mtTest and I reserve with `copy_flag` at 1000-1100, then that range will receive `mtTest`. That's a bit spooky. Perhaps the `copy_flag` should only be used if there's a range starting at exactly that address? That seems less surprising to me.

You are right with this point, but it happens now at call to `uncommit_mapping` where the `copy_flag` is true. `reserve_mapping` always uses `false` for the `copy_flag`. Is this correct if we `assert` that the `leq` is `Committed` or `Reserved`? Is this also correct for commit and `copy_flag`?

File: test/hotspot/gtest/nmt/test_vmatree.cpp
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:178]: Reserving over already reserved memory would replace the old mapping and thus destroy the old content (as does committing, btw). So, this is rather rare. When its not an error, it is done to e.g. change mapping properties before starting to use the mapping.

**Comment by tstuefe** [comment]
@afshin-zafari thank you for implementing this

> > One thing that worries me a bit is the fact that we take the `leq` value without checking if it's equal to the `A` address. This means that if we only have a range 0-100 with mtTest and I reserve with `copy_flag` at 1000-1100, then that range will receive `mtTest`. That's a bit spooky. Perhaps the `copy_flag` should only be used if there's a range starting at exactly that address? That seems less surprising to me.
> 
> You are right with this point, but it happens now at call to `uncommit_mapping` where the `copy_flag` is true. `reserve_mapping` always uses `false` for the `copy_flag`. Is this correct if we `assert` that the `leq` is `Committed` or `Reserved`? Is this also correct for commit and `copy_flag`?

We should not test the address value of A. However, we should check the outgoing state of A and *only copy the flag if outgoing node state is either `Committed` or `Reserved`*. We should require a flag from the caller if outgoing state is `Released`: it means we are reserving a new mapping inside what was before a hole in the address space with no pre-existing mapping. So there is no mapping to take over the flag from.

Some examples, lets say we have: 

`reserve(0, 100, mtNMT);`

Now we do:

A) `reserve(50, 60)` 	(inside) -> should take mtNMT
B) `reserve(0, 20)` 	(same starting address) -> should take mtNMT
C) `reserve(90, 100)` 	(same ending address) -> should take mtNMT

But

D) `reserve(100, 200)`   should require flag from caller: even though its adjacent to the old area, its a new area
E) `reserve(1000, 1100)` should require flag from caller


Side note, a theoretical question is what to do with partly overlapping areas: `reserve(90, 110)`

The current code would take over the flag from the original mapping between 10 and 100. And I think that is fine. We don't expect this to happen in hotspot anyway.

File: src/hotspot/share/nmt/vmatree.cpp
Comment (tstuefe) [inline src/hotspot/share/nmt/vmatree.cpp:37]: We need to think about what the behavior should be if the caller passes `copy_flag=true` but vmatree disagrees and thinks the caller should specify a flag. At the moment we then quietly take over the flag from the `metadata` argument, but I think we should assert, then. Something like "reserving without pre-existing range requires a valid NMT flag".

Tied into that is the question of what the flag in the `metadata` argument should be for `copy_flag=true`. I would argue that in that case, the flag should be `mtNone`. In other words, if I am the caller and I commit memory, I don't want to think about flags. VMATree should figure this out.

In fact, we could argue that `mtNone` means `copy_flag==true` and then we don't need the new `copy_flag` argument. In that case, reserving with `mtNone` means "no clue about the flag, you figure it out" and then it requires a pre-existing mapping. Are there valid use cases for reserving ranges with mtNone as a valid NMT flag?

Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:37]: Asserting may make sense on debug builds, as you might be making a mistake, but we should have pre-defined behavior for production builds anyhow, no `ShouldNotReachHere()`. I think that valid behavior is to set the flag to whichever was provided in the call.


Another option is to have a more elaborate heuristic. If you uncommit `[a, b)` then find the first range which starts inside of `[a, b)` and take its memory flag, if none is found then give up and take `mtNone`.

The nicest behavior is an `O(n)` where we go through all regions in `[a, b)` and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved. 

>Are there valid use cases for reserving ranges with mtNone as a valid NMT flag?

Only in the context of then later setting the memory flag, which I think is a pattern which should be avoided if possible.

Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:37]: I'm only considering the usage of this flag as something which should be used by an `uncommit_region` API, if that wasn't clear.

Comment (tstuefe) [reply src/hotspot/share/nmt/vmatree.cpp:37]: > Asserting may make sense on debug builds, as you might be making a mistake, but we should have pre-defined behavior for production builds anyhow, no ShouldNotReachHere(). I think that valid behavior is to set the flag to whichever was provided in the call.

I disagree :)

The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.

By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.

> Another option is to have a more elaborate heuristic. If you uncommit [a, b) then find the first range which starts inside of [a, b) and take its memory flag, if none is found then give up and take mtNone.

This does not sound very appealing tbh. In most cases, this would be an error. It is not NMTs responsibility to catch this error, but I would not go out of my way to accomodate it case either. 

> The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.

You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?

> I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.

Restricting this to uncommit would not use its full potential.

Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.

The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.

Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:37]: > I disagree :)
> 
> The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.
> 
> By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.

I don't understand what you disagree with specifically. I'm saying that asserting is fine, but we shouldn't give up and kill the VM in release builds with a `ShouldNotReachHere()` in case the programmer messed up when reporting diagnostics to NMT. It's not necessarily a bug in the actual 'business logic' of the code and we can recover from it by providing worse diagnostics.

> 
> 
>     The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.
> 
> You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?

I was thinking in the sense of `uncommit`. Example:

```
reserve(0, 100, mtTest);
reserve(600, 1000, mtTest);
//  [0, 100) and [600, 1000) reserved with flag mtTest
commit(0, 100); commit(600, 1000);
//  [0, 100) and [600, 1000) committed with flag mtTest
// Now we can do
uncommit(0, 1000);
// And we end up with [0, 100) and [600, 1000) being reserved with flag mtTest
```

That is not equivalent to `register_mapping` right now, but `register_mapping` already does everything to support this kind of operation.




> 
>     I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.
> 
> Restricting this to uncommit would not use its full potential.
> 
> Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.
> 
> The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.

I don't disagree with this, I wasn't thinking about committing when I wrote my comments.

Comment (tstuefe) [reply src/hotspot/share/nmt/vmatree.cpp:37]: > > I disagree :)
> > The point of this default mechanism is to take cognitive load from the programmer. I don't want to have to think about the flag when I rely on this mechanism. Having this mechanism, but then still forcing the caller to specify a flag "just in case you commit into unreserved territory" is not a big help.
> > By using copy_flag=true, I declare that I expect there to be a prior mapping in that range. That is part of the API contract. If that expectation is broken, I did something wrong. This is not something that can happen due to random runtime conditions. Therefore, like any other invalid state it should be handled with an assert. I don't think it makes a lot of sense to think about what to do in release builds: this is an error like many others, and we have many asserts in the JVM and we only rarely encode a valid non-assert path.
> 
> I don't understand what you disagree with specifically. I'm saying that asserting is fine, but we shouldn't give up and kill the VM in release builds with a `ShouldNotReachHere()` in case the programmer messed up when reporting diagnostics to NMT. It's not necessarily a bug in the actual 'business logic' of the code and we can recover from it by providing worse diagnostics.
> 
> > ```
> > The nicest behavior is an O(n) algorithm where we go through all regions in [a, b) and convert each to its reserved version with the flag, then uncommit with holes in the range would leave those holes unreserved.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You lost me here. This is what we do right now, isn't it? VMATree just replaces all existing mappings between [a, b) with the new mapping?
> 
> I was thinking in the sense of `uncommit`. Example:
> 
> ```
> reserve(0, 100, mtTest);
> reserve(600, 1000, mtTest);
> //  [0, 100) and [600, 1000) reserved with flag mtTest
> commit(0, 100); commit(600, 1000);
> //  [0, 100) and [600, 1000) committed with flag mtTest
> // Now we can do
> uncommit(0, 1000);
> // And we end up with [0, 100) and [600, 1000) being reserved with flag mtTest
> ```

In VMAtree sure, but in reality not: since NMT has no complete knowledge about the address space. Uncommitting a range containing what NMT sees as address holes will wipe any mappings that happen to be in these holes that NMT does not know. Which are quite a lot.

> 
> That is not equivalent to `register_mapping` right now, but `register_mapping` already does everything to support this kind of operation.
> 
> > ```
> > I'm only considering the usage of this flag as something which should be used by an uncommit_region API, if that wasn't clear.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Restricting this to uncommit would not use its full potential.
> > Consider this example: In the Metaspace allocator I would like to commit memory, but that memory can be tagged with either mtClassSpace or mtMetaspace, and in the course of Lilliput we may see even more tags. In all cases I reuse the same allocator code. Here, why should I be forced to have to specify the tag on commit? I already marked the underlying memory with the correct tag when reserving it.
> > The ability of painting regions with a certain flag and then having subsequent memory operations on that memory preserve the flag is powerful. You can, for example, in Metaspace account memory at the chunk level. Without having to add code everywhere that keeps track of those flags. You only have to paint the chunks you give to a Metaspace arena with the correct flag, and from then on the whole process is automatic.
> 
> I don't disagree with this, I wasn't thinking about committing when I wrote my comments.

**Comment by afshin-zafari** [comment]
To avoid confusion and have a reference in further discussions, I made a table that enumerates the possible cases and used for handling the invalid cases.
In SimpleTracer test, the reserve/commit/release operations can happen at any order. When the order is not valid, a warning is shown and the test continues. In a non-tests codes, we want these cases taken as fatal. More discussions needed.

**Comment by afshin-zafari** [comment]
Keep it still alive, bot.

File: src/hotspot/share/nmt/vmatree.hpp
Comment (gerard-ziemski) [inline src/hotspot/share/nmt/vmatree.hpp:192]: In

`SummaryDiff register_mapping(position A, position B, StateType state, const RegionData& metadata, bool copy_flag = false);`

`position B` is clearly a position of its own, however, in

`SummaryDiff reserve_mapping(position from, position sz, const RegionData& metadata)`

`position sz` is **size** or **delta**. We should rename the parameter to reflect that intended functionality better.

I would probably prefer to see us introduce:

`using position_size = size_t;`

and use as appropriate, instead, of forcing everything to be `position`, to help keep the API usage straight.

File: test/hotspot/gtest/nmt/test_vmatree.cpp
Comment (gerard-ziemski) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:179]: OK thank you, just wanted to make sure that this usage is what we want and allow. I initially thought we would not want that and have some assert checking for it.

File: test/hotspot/gtest/nmt/test_vmatree.cpp
Comment (gerard-ziemski) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:195]: Is there a way not to need this casting?

`(size_t)node->key()`

File: src/hotspot/share/nmt/vmatree.cpp
Comment (gerard-ziemski) [inline src/hotspot/share/nmt/vmatree.cpp:107]: Can we replace hardcoded `-1` with some a constant in:

`VMATree::SummaryDiff invalid_diff(-1);`

and what is the caller supposed to do with that? How would it check for the special INVALID_DIFF value?

For example, in `src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp` we define:

`constexpr static const StackIndex invalid = std::numeric_limits<StackIndex>::max() - 1;`

and offer the API:

`static bool is_invalid(StackIndex a)`

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

File: src/hotspot/share/nmt/vmatree.hpp
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.hpp:192]: Hi, I did this change in https://github.com/openjdk/jdk/pull/20994

**Comment by jdksjolen** [comment]
I'd like to suggest the following change set as the PR: https://github.com/openjdk/jdk/compare/master...jdksjolen:jdk:_8335091_vmatree_copy_mtflag_on_commit

**Comment by afshin-zafari** [comment]
> I'd like to suggest the following change set as the PR: [master...jdksjolen:jdk:_8335091_vmatree_copy_mtflag_on_commit](https://github.com/openjdk/jdk/compare/master...jdksjolen:jdk:_8335091_vmatree_copy_mtflag_on_commit)
> 
> It adds a set of `assert`s for when we believe that the supplied `copy_flag` argument is incorrect and has reasonable defaults.

All the changes are applied now.

**Comment by afshin-zafari** [comment]
The conflicts come from the PR for renaming MEMFLAGS to MemTag.

**Comment by jdksjolen** [comment]
These changes look good to me, but we should probably get a thumbs up from @tstuefe and @gerard-ziemski before merging. Also, need to include the UL headers to fix the compilation issue.

File: src/hotspot/share/nmt/vmatree.cpp
Comment (gerard-ziemski) [reply src/hotspot/share/nmt/vmatree.cpp:38]: The argument name `copy_flag` suggests to me that we will be using the provided `flag` from `metadata`, but the implementation does the reverse.

I think we might need a better name, ex: `use_implicit_flag` in which case we copy the flag out of the region itself, or we could use `use_explicit_flag` and do the opposite.

File: src/hotspot/share/nmt/vmatree.hpp
Comment (gerard-ziemski) [inline src/hotspot/share/nmt/vmatree.hpp:184]: Why do we need this API? It's only used in the test so far, so do we really need it?

I am not too crazy about `register_mapping` basically being a wrapper around `register_mapping(.., true)`

What if someone calls `commit_mapping(.., true)` ? Doesn't that turn into `register_mapping(..., StateType::Committed, metadata, true)` which is the same as `uncommit_mapping()` ? It doesn't make sense?

File: src/hotspot/share/nmt/vmatree.hpp
Comment (jdksjolen) [inline src/hotspot/share/nmt/vmatree.hpp:187]: As Gerard said: This should be `StateType::Reserved`, clearly this isn't caught by our tests so we do need a test that catches this.

File: src/hotspot/share/nmt/vmatree.cpp
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:38]: I see what you're saying, maybe `use_found_flag` or `use_flag_inplace`? At least the assertions give a strong hint that that assumption is wrong. Better names are (almost) always welcome.

File: src/hotspot/share/nmt/vmatree.hpp
Comment (afshin-zafari) [reply src/hotspot/share/nmt/vmatree.hpp]: `StateType::Reserved` is used instead.

File: src/hotspot/share/nmt/vmatree.hpp
Comment (afshin-zafari) [reply src/hotspot/share/nmt/vmatree.hpp]: New check is added to `NMTVMATreeTest.UseFlagInplace` gTest.

**Comment by jdksjolen** [comment]
One thing: We need to merge this with master and update the naming from "flag" to "tag" to be compliant with the new naming scheme.

**Review by jdksjolen — APPROVED** [review]
I'm OK with this as long as the flag -> tag change is made also.

**Review by gerard-ziemski — APPROVED** [review]
LGTM assuming we merge with latest and it passes the tests.

**Review by jdksjolen — APPROVED** [review]

**Comment by afshin-zafari** [comment]
flag -> tag is done.
merge is done.
All tier1 tests passed locally (linux-x64).
GHA tests all passed.

Dear @tstuefe , if you don't have any comment I can integrate this PR.

**Review by gerard-ziemski — APPROVED** [review]
LGTM

**Comment by jdksjolen** [comment]
If this passes tests then I think it should be integrated

**Pull request closed without merging**
